Keycloak How-to (#1839)
spectrocloud/librarium: docs/docs-content/user-management/saml-sso/palette-sso-with-keycloak.md

## Setup

1. Ensure you can access your Kubernetes cluster using the `kubectl` CLI. Refer to the [Access Cluster with CLI](../../clusters/cluster-management/palette-webctl.md) for guidance on how to access your cluster with the `kubectl` CLI.

2. Install Keycloak on your Kubernetes cluster using the default configuration. This will create a Keycloak deployment and a se rvice with a LoadBalancer type. The service will be exposed on an external IP address.

   ```bash
   kubectl create --filename https://raw.githubusercontent.com/keycloak/keycloak-quickstarts/latest/kubernetes/keycloak.yaml
   ```

   :::note

   You can also download the YAML file and edit it to suit your requirements before deploying it to your cluster.

   :::

3. The install will take a couple of minutes. Use the following command to retrieve the external IP address of the Keycloak service.

  ```bash
  kubectl describe service keycloak  | grep "LoadBalancer Ingress" | awk '{print $3}' && \
  IP=$(kubectl describe service keycloak | grep "LoadBalancer Ingress" | awk '{print $3}')
  ```

  ```shell hideClipboard
  aacf4014d5cd34825803567201217410-1398304919.us-east-1.elb.amazonaws.com
  ```

> 4. Next, download the Ingress YAML definition provided by Keycloak to create an Ingress resource in your cluster. The command below will also replace the `KEYCLOAK_HOST` placeholder with the external IP address of the Keycloak service.

Reviewer: ```suggestion
4. Next, download the Ingress YAML definition provided by Keycloak to create an Ingress resource in your cluster. The command below also automatically replaces the `KEYCLOAK_HOST` placeholder with the external IP address of the Keycloak service.
```

  ```bash
  wget --quiet --output-document - https://raw.githubusercontent.com/keycloak/keycloak-quickstarts/latest/kubernetes/keycloak-ingress.yaml | sed "s/KEYCLOAK_HOST/$IP/" | kubectl create -f -
  ```

  ```shell hideClipboard
  ingress.networking.k8s.io/keycloak created
  ```

>   :::note

Reviewer: ```suggestion
  :::info
```

>   If `wget` and `sed` are not available, download the file and manually edit the file replacing KEYCLOAK_HOST with the external IP address of the Keycloak service.

Reviewer: ```suggestion
  If `wget` and `sed` are not available, download the file and manually edit the file to replace `KEYCLOAK_HOST` with the external IP address of the Keycloak service.
```

  :::

>   Upon successful installation, the following services will be available in your cluster. You can review the exposed services in the cluster details page.

Reviewer: ```suggestion
  After the ingress resource is created, the following services will be available in your cluster. You can review the exposed services in the cluster details page.
```

  ![alt_text](/keycloak/user-management_saml-sso_keycloak-01-keycloak-service.png "Keycloak Service")

> 5. Create a DNS CNAME for the URL exposed by load balancer. For example, the CNAME `keycloak.dmitry.sa.spectrodemos.com` points to the following URL exposed by the load balancer `aacf4014d5cd34825803567201217410-1398304919.us-east-1.elb.amazonaws.com`. 

Reviewer: ```suggestion
5. Create a DNS CNAME record for the URL exposed by load balancer. For example, the CNAME `keycloak.dmitry.sa.spectrodemos.com` points to the following URL exposed by the load balancer `aacf4014d5cd34825803567201217410-1398304919.us-east-1.elb.amazonaws.com`. 
```

> 6. Login to Keycloak with by using the domain name you created in the previous step, or you can use the exposed load balancer URL. The Keycloak admin console is available on port `8080/admin`. For example, `http://keycloak.dmitry.sa.spectrodemos.com:8080/admin`. Use the default credentials `admin:admin` to login to the admin console.

Reviewer: ```suggestion
6. Log in to Keycloak with by using the domain name you created in the previous step, or you can use the exposed load balancer URL. The Keycloak admin console is available on port `8080/admin`. For example, `http://keycloak.dmitry.sa.spectrodemos.com:8080/admin`. Use the default credentials `admin:admin` to log into the admin console.
```

  :::caution

>   We recommend that you change the default admin password after logging in to the admin console.

Reviewer: ```suggestion
  We recommend that you change the default credentials after logging in to the admin console.
```

  :::

  ![Keycloak Admin console](/keycloak//user-management_palette-rback_keycloak_login.png)

> 7.  Next, login to [Palette](https://console.spectrocloud.com), and navigate to the left **Main Menu** and select **Tenant Settings**. Next, select **SSO** from the **Tenant Menu** to access the SSO configuration page. Click on the **OIDC** tab to configure OIDC for Palette. Copy the values in the **Callback URL** and **Logout URL** fields. You will need these values to configure Keycloak.

Reviewer: ```suggestion
7.  Next, log in to [Palette](https://console.spectrocloud.com), and navigate to the left **Main Menu** and select **Tenant Settings**. Next, select **SSO** from the **Tenant Menu** to access the SSO configuration page. Click on the **OIDC** tab to configure OIDC for Palette. Copy the values in the **Callback URL** and **Logout URL** fields. You will need these values to configure Keycloak.
```

  ![alt_text](/keycloak/user-management_saml-sso_keycloak-02-callback-url.png "Callback URL")

8. Switch back to the Keycloak admin console and create a client for Palette. Navigate to the left **Main Menu** and select **Clients**. Click on the **Create** button to create a new client.

  ![alt_text](/keycloak/user-management_saml-sso_keycloak-03-create-client.png "Create Client")

9. Select **Client Type** value "OpenID Connect" and fill in the fields **Client ID** and **Name** with the value `palette`. Click on **Next**.

  ![alt_text](/keycloak/user-management_saml-sso_keycloak-04-palette-username.png "Palette username")

10. Select **Client authentication** for increased security and check the **Standard Flow** box and the **Direct Access Grants** box. Click on **Next**.

  ![alt_text](/keycloak/user-management_saml-sso_keycloak-05-client-authentication.png "Client Authentication")

> 11. Fill out the following fields. Refer to the table below for more information.

Reviewer: ```suggestion
11. Fill out the following fields with the instructions provided in the table.
```

  | **Field** | **Description** |
  | --- | --- |
  | **Root URL** | Your Palette URL. For Palette SaaS, this value is prefixed with your tenant name. For example, `https://docs-test.console.spectrocloud.com`. For self-hosted Palette, or VerteX, this value is the URL of your Palette deployment. |
  | **Valid Redirect URIs** | The callback URL you copied from the Palette SSO configuration page. |
  | **Valid post logout redirect URIs** | The logout URL you copied from the Palette SSO configuration page. |  

12. Click on **Save** to save the client configuration.

  ![alt_text](/keycloak/user-management_saml-sso_keycloak-06-keycloak-callback.png "Keycloak callback")

13. In the following screen, select the **Credentials** tab to retrieve client secret. 

  ![alt_text](/keycloak/user-management_saml-sso_keycloak-07-keycloak-credentials.png "Keycloak Credentials")

> 14. Switch back to Palette and paste client secret in the **Client Secret** field. Fill out the following fields. Refer to the table below for more information.

Reviewer: ```suggestion
14. Switch back to Palette and paste client secret in the **Client Secret** field. Fill out the following fields with the instructions provided in the table below.
```
